k2/csrc/fsa.h

```diff
@@ -205,8 +205,40 @@ std::ostream &operator<<(std::ostream &os, const DenseFsaVec &dfsavec);
 */
 Fsa FsaFromTensor(Tensor &t, bool *error);
 
+/*
+  Create an FSA from a Array1. The Array1 is expected to has N arcs (the format
+  is src_state, dest_state, symbol, cost). The cost is not really an int32_t, it
+  is a float. This code will print an error message and output 'true' to
+  'error', and return an empty FSA (with no states or arcs) if 't' was not
+  interpretable as a valid FSA. These requirements for a valid FSA are:
+    - src_state values on the arcs must be non-decreasing
+    - all arcs with -1 as the label must be to a single state (call this
+      final_state) which has no arcs leaving it
+    - final_state, if it exists, must be numerically greater than any state
+      which has arcs leaving it, and >= any state that has arcs entering it.
+  If there are no arcs with -1 on the label, here is how we determine the final
+  state:
+     - If there were no states at all in the FSA we'll return the empty FSA
+       (with no arcs), and set error = true. (As the k2-fsa definition is there
+       at least must be a super final state.)
+     - Else if there were some states but no arcs, we'll return the FSA (with
+       just one implicit super final state (set id = 1) and no arcs), and set error = false.
+     - Otherwise, we'll let `final_state` be the highest-numbered state
+       that has any arcs leaving or entering it, plus one.  (This FSA
+       has no successful paths but still has states.)
 
-Fsa FsaFromArray1(Array1<Arc> &arc, bool *error);
+    @param [in] array       Source arcs. Caution: the returned FSA will share
+                            memory with this tensor, so don't modify it
+                            afterward!
+    @param [in] num_states  Here, the num_states means the number of all states
```

Comment:
Let's change this interface so the id of the final-state is provided, instead of the num_states.  THen there will be no confusion about whether the states are contiguous.